cmd/stdiscosrv: New discovery server (ref #4618)

calmh opened the pull request:
### Purpose

This is a new revision of the discovery server. Relevant changes and non-changes:

- Protocol towards clients is unchanged.

- Recommended large scale design is still to be deployed nehind nginx (I
  tested, and it's still a lot faster at terminating TLS).

- Database backend is leveldb again, only. It scales enough, is easy to
  setup, and we don't need any backend to take care of.

- Server supports replication. This is a simple TCP channel - protect it
  with a firewall when deploying over the internet. (We deploy this within
  the same datacenter, and with firewall.) Any incoming client announces
  are sent over the replication channel(s) to other peer discosrvs.
  Incoming replication changes are applied to the database as if they came
  from clients, but without the TLS/certificate overhead.

- Metrics are exposed using the prometheus library, when enabled.

- The database values and replication protocol is protobuf, because JSON
  was quite CPU intensive when I tried that and benchmarked it.

- The "Retry-After" value for failed lookups gets slowly increased from
  a default of 120 seconds, by 5 seconds for each failed lookup,
  independently by each discosrv. This lowers the query load over time for
  clients that are never seen. The Retry-After maxes out at 3600 after a
  couple of weeks of this increase. The number of failed lookups is
  stored in the database, now and then (avoiding making each lookup a
  database put).

All in all this means clients can be pointed towards a cluster using
just multiple A / AAAA records to gain both load sharing and redundancy
(if one is down, clients will talk to the remaining ones).

### Testing

It's currently live at discovery.s.n / discovery-v4.s.n / discovery-v6.s.n. There is a monitoring dashboard at https://monitor.syncthing.net/ based on the new prometheus metrics.

Clients are currently using it by way of discovery-v4-4.s.n and discovery-v6-4.s.n also being pointed towards these boxes, which is intended to remain the case for a long time.

 ### Reviewing

I don't think the diffs make too much sense any more. It's probably reasonable to just review the apisrv (previously "querysrv"), database and replication layers by themselves. The rest of the mechanics is the same.

There are some vendor updates for the new prometheus packages, removing the old database packages, and adding the protobuf package (was missing) for the compilation of the new structs.

AudriusButkevicius commented:
Catching a flight now. Will come back to this.

calmh commented:
No hurry on this at all, I just wanted it up here for visibility. Fly safely.

AudriusButkevicius commented on cmd/stdiscosrv/replication.go:
```diff
@@ -0,0 +1,202 @@
+)
+
+type replicator interface {
+	send(key string, ps []DatabaseAddress, seen int64)
```
Wtf ps

calmh replied:
Renamed

AudriusButkevicius commented on cmd/stdiscosrv/replication.go:
```diff
@@ -0,0 +1,202 @@
+	case s.inbox <- item:
+	default:
+		replicationSendsTotal.WithLabelValues("drop").Inc()
+	}
```
This is shit, our lookups become probabilistic

calmh replied:
They've always been, but in practice this never happens. Nonetheless it's the right thing to do because we don't want to block any local operations due to a stalled replication

AudriusButkevicius replied:
We should have some sort of periodic reconcile in that case?

AudriusButkevicius replied:
Nvm read the comment below.

AudriusButkevicius commented:
If we restart one of these things, the picture can become inconsistent.

calmh commented on cmd/stdiscosrv/apisrv.go:
```diff
@@ -0,0 +1,383 @@
+		// some sort of internal error
+		w.Header().Set("Retry-After", errorRetryAfterSeconds)
+		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		return
```
Added

AudriusButkevicius replied:
Still nothing in this PR

calmh replied:
It's in the db layer

calmh replied:
And the API fail gets sort of automatically counted as well as the top level handler picks up the 500 status code and that's a statistic

AudriusButkevicius replied:
Well this is very confusing as the if above has special handling and metric logging?

calmh replied:
Oh I see what you mean now. Yeah we could add a result type for lookupRequestsTotal here, I'd missed that.

calmh commented on cmd/stdiscosrv/apisrv.go:
```diff
@@ -0,0 +1,383 @@
+
+	lookupRequestsTotal.WithLabelValues("success").Inc()
+
+	bs, _ := json.Marshal(announcement{Addresses: addresses(rec.Addresses)})
```
If there are no addresses we return an empty 404, and (occasionally) clean up the db. Note that we never delete keys from the database, but the value will in time be replaced by something with nil addresses, a seen date in the (distant) past and a slowly increasing misses counter if people are still asking for it.

AudriusButkevicius replied:
Applying the addresses function earlier is still relevant.

calmh replied:
Why? It isn't used until this point.

AudriusButkevicius replied:
Why is no addresses known not the same effect as no unexpired addresses known?

calmh replied:
You're confusing one with the other. Expired addresses get filtered out already at the db get level, this is just converting DatabaseAddress structs to the string with the address

AudriusButkevicius replied:
Right, so let's not do filtering there then, as it's confusing. Or add a comment explaining why we are doing filtering given the result should already be filtered?

calmh replied:
There is no filtering here so I'm not sure what you mean. The magic in the db is that it expires addresses. I think that fits at that layer. When doing this against S3, etcd, dynamodb and other "real" kv-stores the expire was handled by them, when we did this with postgresql it was handled in the WHERE clause, now it's handled in the get. I fail to see the problem.

calmh replied:
Ha! I was wrong of course, which explains the confusion. There *should* be no filtering here. Removing.

calmh commented on cmd/stdiscosrv/apisrv.go:
```diff
@@ -0,0 +1,383 @@
+	now := time.Now()
+	expire := now.Add(addressExpiryTime).UnixNano()
+
+	ps := make([]DatabaseAddress, len(addresses))
```
A name from when the type was called something else. Renaming.

calmh commented on cmd/stdiscosrv/apisrv.go:
```diff
@@ -0,0 +1,383 @@
+	seen := now.UnixNano()
+	if s.repl != nil {
+		s.repl.send(key, ps, seen)
+	}
```
It's nonblocking, as opposed to the database put

calmh commented on cmd/stdiscosrv/apisrv.go:
```diff
@@ -0,0 +1,383 @@
+		}
+
+		uri.Host = net.JoinHostPort(host, port)
+		fixed = append(fixed, uri.String())
```
Fixed

AudriusButkevicius replied:
Still appends instead of addresses[i] = url.String()

calmh replied:
No reallocation though as the slice is already made. This is a valid pattern... But I can super clarify it like that, sure

calmh replied:
Actually no because some addresses might get filtered out, this is the clearest way of expressing it and maintaining few allocations

calmh commented on cmd/stdiscosrv/apisrv.go:
```diff
@@ -0,0 +1,383 @@
+}
+
+func addresses(ps []DatabaseAddress) []string {
+	res := make([]string, 0, len(ps))
```
Renamed, and yeah as the element type is different, unless you mean an `addressesInto([]string, ...)` with some sort of reused buffer but that seems overkill to me

calmh commented on cmd/stdiscosrv/database.go:
```diff
@@ -0,0 +1,264 @@
+}
+
+type database interface {
+	put(key string, rec DatabaseRecord) error
```
Hm. It sort of depends on what comes in. If there are no addresses in the record it becomes a straight put. How about `save()`?

AudriusButkevicius replied:
No, it took me ages to realize what happens to the non-expired old entries.

calmh replied:
How about I split it into two separate functions:

```
put(rec DatabaseRecord) // straight put, no magic
merge(addrs []DatabaseAddress) // with magic
```

calmh replied:
Replication send is more like merge and can have the same signature

AudriusButkevicius replied:
Fine.

calmh commented on cmd/stdiscosrv/database.go:
```diff
@@ -0,0 +1,264 @@
+	t0 := time.Now()
+	rc := make(chan error)
+	s.inbox <- func() {
+		if len(rec.Addresses) != 0 {
```
Nope, as we also do `put(DatabaseRecord{Misses: 42, Seen: alongTimeAgo})` to keep track of devices that are unknown but being asked about

calmh commented on cmd/stdiscosrv/database.go:
```diff
@@ -0,0 +1,264 @@
+			}
+		}
+		size := rec.Size()
+		if len(s.marshalBuf) < size {
```
It's done that way because the slice is never resliced, as MarshalTo just overwrites the first part of it, cap is always == len and len is thus less magical

calmh commented on cmd/stdiscosrv/database.go:
```diff
@@ -0,0 +1,264 @@
+			// failure if there's something wrong with it.
+			var rec DatabaseRecord
+			if err := rec.Unmarshal(iter.Value()); err != nil {
+				s.db.Delete(iter.Key(), nil)
```
This is actually a "can't happen" condition, should probably be removed entirely rather than optimized. (It kicked in once when I changed serialization from JSON to proto...)

calmh commented:
Yeah, or if it's down for a significant amount of time. I toyed with the idea of sending the whole database on replication connect, but it didn't seem worth it. Any inconsistency will be sorted out in at most an hour, and in the meantime an incorrect negative answer will result in a new question a couple of minutes later with a 50% or 66% chance of getting an answer, at which point it's resolved. I think this is a case where 99% correct is correct enough.

(The old disco clusters are dropping light 70% of all queries at the moment...)

calmh commented:
I don't want to build one more eventually consistent multimaster distributed kv-store that sucks, I just want something that mostly works for our purposes. :)

AudriusButkevicius commented on cmd/stdiscosrv/database.go:
```diff
@@ -0,0 +1,311 @@
+			if newRec.Seen < oldRec.Seen {
+				newRec.Seen = oldRec.Seen
+			}
+		}
```
It feels as if merge should take two recs and return a merged rec, rather than merge addresses and then leave you to write extra code to merge the rest.

calmh replied:
Yeah, correct...

AudriusButkevicius commented on cmd/stdiscosrv/database.go:
```diff
@@ -0,0 +1,311 @@
+}
+
+// expire returns the list of addresses after removing expired entries
+func expire(addrs []DatabaseAddress, now int64) []DatabaseAddress {
```
Again, in place delete, like in other places would be better, but this is nitpicking. Gotta shave off those uS for save costs on infra.

AudriusButkevicius commented:
@st-review lgtm

st-review commented:
@AudriusButkevicius: Noted! Need another LGTM or explicit merge command.

AudriusButkevicius commented:
Merge once the builds pass I guess

calmh commented:
I'm going to do one more thing, which is implement TLS on the replication channel. Because I figure we might as well spread the cluster over a couple of european cities, the extra millisecond of latency doesn't matter, and then it's irresponsible to run in clear text and without auth.

My plan is to use the "usual" method with device IDs matching the certificate. Each peer gets a device ID in the `--replication-send` flag, and the listener will accept connections from the same device IDs mentioned there.

AudriusButkevicius commented:
Makes sense.

AudriusButkevicius commented on cmd/stdiscosrv/replication.go:
```diff
@@ -0,0 +1,292 @@
+		if !deviceIDIn(remoteID, l.allowedIDs) {
+			log.Println("Replication accept: unexpected device ID:", remoteID)
+			continue
+		}
```
Leaks connection in a few places.

AudriusButkevicius commented:
Also, too many command line options. I'd go with -replicate=devid@host:port,... Or even just host, and having a default port for replication which is listened on if any repl device is specified.

calmh commented:
Updated, docs in https://github.com/syncthing/docs/pull/369

AudriusButkevicius commented:
A bit of my own medicine I guess. The forced push lost the history of what was reviewed. What has changed?

Also, did you try to add some 'dump everything I have on connect' to reconcile after restarts?

calmh commented:
Changes since last are fixing the connection leaks you pointed out, cleaner command line syntax for replication IDs, and depending on when you reviewed, adding TLS at all to the replication. Nothing done about sending all att connect, although I bumped the channel size to 10.000 (named constant too). With our rates (~60 sends/s) that buffers about three minutes of sends so handles a short reboot. (Apart from the messages in flight or in network buffers when the shutdown happens, which are lost.)

AudriusButkevicius commented:
Looks good then I guess.

calmh commented:
@st-review merge it then

st-review commented:
:ok_hand: Merged as 916ec63af684429955cb8c7c4fbe81d684f9c4ee. Thanks, @calmh!
